File: adapters/src/main/java/com/wang/container/interfaces/IItemClick.java
package com.wang.container.interfaces;

import android.view.View;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;

import com.wang.container.R;
import com.wang.container.holder.BaseViewHolder;

/**
 * OnItemClickListener的接口
 * 见子类实现{@link OnItemClickListener}
 */
public interface IItemClick extends View.OnClickListener, View.OnLongClickListener {

    @Override
    @CallSuper//一般不需要重写，所以加了此限制（如果真的不想调用super可以注解抑制掉错误）
    default void onClick(@NonNull View view) {
        onItemClick(view, getViewPosition(view));
    }

    @Override
    @CallSuper
    default boolean onLongClick(@NonNull View view) {
        return onItemLongClick(view, getViewPosition(view));
    }

    /**
     * 获取当前view所在的position
     */
    int getViewPosition(@NonNull View view);

    /**
     * 获取当前view所在的ViewHolder
     */
    @CallSuper
    default BaseViewHolder getViewHolder(@NonNull View view) {
        return (BaseViewHolder) view.getTag(R.id.tag_view_holder);
    }

    /**
     * 获取当前view所在的adapter
     */
    @CallSuper
    default IAdapter getAdapter(@NonNull View view) {
        return (IAdapter) view.getTag(R.id.tag_view_adapter);
    }

    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // 回调方法
    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /**
     * item被点击时
     *
     * @param position 属于该adapter的position
     */
    void onItemClick(@NonNull View view, int position);

    /**
     * item被长按时
     */
    default boolean onItemLongClick(@NonNull View view, int position) {
        return false;
    }
}

File: adapters/src/main/java/com/wang/container/interfaces/IItemClick.kt
package com.wang.container.interfaces

import android.view.View
import android.view.View.OnLongClickListener
import androidx.annotation.CallSuper
import com.sea.base.adapter.BaseViewHolder
import com.sea.im.base.R

/**
 * OnItemClickListener的接口
 * 见子类实现[OnItemClickListener]
 */
interface IItemClick : View.OnClickListener, OnLongClickListener {

    /**
     * 获取当前view所在的position
     */
    fun getViewPosition(view: View): Int

    /**
     * 获取当前view所在的ViewHolder
     */
    @CallSuper
    fun getViewHolder(view: View): BaseViewHolder<*> {
        return view.getTag(R.id.tag_view_holder) as BaseViewHolder<*>
    }

    /**
     * 获取当前view点击时的tag
     */
    @CallSuper
    fun getViewClickTag(view: View): String? {
        return view.getTag(R.id.tag_view_adapter_item_view_tag) as? String
    }
    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // 回调方法
    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    /**
     * item被点击时
     *
     * @param position 属于该adapter的position
     */
    fun onItemClick(view: View, position: Int)

    /**
     * item被长按时
     */
    fun onItemLongClick(view: View, position: Int): Boolean

    /**
     * item里的view被点击时
     *
     * @param position 属于该adapter的position
     */
    fun onItemViewClickWithTag(view: View, position: Int, tag: String)

    /**
     * item里的view被长按时
     */
    fun onItemViewLongClickWithTag(view: View, position: Int, tag: String): Boolean
}